src/safeposix/syscalls/fs_calls.rs

```diff
@@ -1537,4 +1537,80 @@ impl Cage {
             }
         }
     }
+    
+    //------------------GETDENTS SYSCALL------------------
+
+    pub fn getdents_syscall(&self, fd: i32, dirp: *mut u8, bufsize: usize)-> i32 {
+        
+        let mut vec: Vec<(interface::ClippedDirent, Vec<u8>)> = Vec::new();
+
+        // make sure bufsize is at least greater than size of a ClippedDirent struct
+        if bufsize <= interface::CLIPPED_DIRENT_SIZE {
+            return syscall_error(Errno::EINVAL, "getdents", "Result buffer is too small.");
+        }
+        
+        let fdtable = self.filedescriptortable.read().unwrap();
+
+        if let Some(wrappedfd) = fdtable.get(&fd) { // check if fd is valid
+            let mut filedesc_enum = wrappedfd.write().unwrap();
+            
+            match &mut *filedesc_enum {
+                // only proceed when fd represents a file
+                File(ref mut normalfile_filedesc_obj) => {
+                    let metadata = FS_METADATA.read().unwrap();
+                    let inodeobj = metadata.inodetable.get(&normalfile_filedesc_obj.inode).unwrap();
+
+                    match inodeobj {
+                        // only proceed when inode is a dir
+                        Inode::Dir(dir_inode_obj) => {
+                            let position = normalfile_filedesc_obj.position;
+                            let mut bufcount = 0;
+                            let mut curr_size = 0;
+                            let mut count = 0;
+                            let mut temp_len = 0;
+
+                            // iterate over filename-inode pairs in dict
+                            for (filename, inode) in dir_inode_obj.filename_to_inode_dict.clone().into_iter() {
+                                // convert filename to a filename vector of u8
+                                let mut vec_filename: Vec<u8> = filename.as_bytes().to_vec();
```

Comment:
The filename is not null terminated, so we will need to convert between the rust format and the C format. A very hacky way to do this for now would just be to push a null byte at the end, but that's not entirely correct so you can implement that but put an explanatory comment. Alternatively you could put `Cstring::new` into the interface, and use that here.